File: src/main/java/komu/blunt/asm/opcodes/OpPushRegister.java
package komu.blunt.asm.opcodes;

import komu.blunt.asm.Register;
import komu.blunt.asm.VM;

import static com.google.common.base.Preconditions.checkNotNull;

public class OpPushRegister extends OpCode {

    private final Register register;

    public OpPushRegister(Register register) {
        this.register = checkNotNull(register);
    }

    @Override
    public void execute(VM vm) {
        vm.push(vm.get(register));
    }

    @Override
    public boolean modifies(Register register) {
        return false;
    }

    @Override
    public String toString() {
        return String.format("(push %s)", register);
    }
}

File: src/main/kotlin/komu/blunt/asm/opcodes/OpPushRegister.kt
package komu.blunt.asm.opcodes

import komu.blunt.asm.Register
import komu.blunt.asm.VM

class OpPushRegister(private val register: Register) : OpCode() {

    override fun execute(vm: VM) {
        vm.push(vm[register])
    }

    override fun modifies(register: Register) = false
    override fun toString() = "push $register"
}
